File: Dock/Globals/Classes/Logger.js
const crypto = require("crypto");

const { logLevel } = require("../Enumerations/LogLevel");
const { logCategory } = require("../Enumerations/LogCategory");
const { logServiceOrigin } = require("../Enumerations/LogServiceOrigin");
const LogFormatter = require("./Logging/LogFormatter");
const LogIngester = require("./Logging/LogIngester");

/**
 * Logger — the single structured logging entry point for Dock. Every call builds
 * a canonical log entry (id, level, category, title, message, service, optional
 * accountId / errorCode / errorReason / additionalData) with the timestamp stamped
 * INTERNALLY (never a caller parameter), hands it to LogIngester for durable
 * persistence into the logEvents collection, and mirrors it to standard output so
 * an operator tailing the service (journald) always has a live trace.
 *
 * The date/time is never a parameter. Level is chosen by the method
 * (debug/info/warning/error); category is a LogCategory enum; title is a string
 * (use LogTitles.* constants for the fixed events); options carry the structured
 * extras. This runs in ALL environments — production included — unlike the old
 * --debug-only file logger it replaces.
 *
 * The legacy log(message, source) and logWorker(...) signatures are preserved so
 * existing call sites keep working, and initialize()/isEnabled()/getSessionId()
 * retain their meaning for the worker-supervision code that reads them.
 */
class Logger
{
    static #debugEnabled = false;
    static #sessionId = null;
    static #environment = "production";
    static #sequenceCounter = 0;

    static initialize()
    {
        Logger.#debugEnabled = process.argv.includes("--debug") || process.env.DOCK_DEBUG === "1";
        Logger.#sessionId = Date.now().toString();
        Logger.#environment = Logger.#resolveEnvironment();
    }

    static #resolveEnvironment()
    {
        const explicitEnvironmentFlag = process.argv.find(argument => argument.startsWith("--environment="));
        if (explicitEnvironmentFlag)
        {
            return explicitEnvironmentFlag.slice("--environment=".length);
        }
        if (process.env.MINDMELD_ENVIRONMENT)
        {
            return process.env.MINDMELD_ENVIRONMENT;
        }
        if (process.argv.includes("--debug"))
        {
            return "local";
        }
        return "production";
    }

    static isEnabled()
    {
        return Logger.#debugEnabled;
    }

    static getSessionId()
    {
        return Logger.#sessionId;
    }

    static #nextSequence()
    {
        Logger.#sequenceCounter = (Logger.#sequenceCounter + 1) % Number.MAX_SAFE_INTEGER;
        return Logger.#sequenceCounter;
    }

    static #record(level, category, title, message, options = {})
    {
        const now = new Date();
        const logEntryDocument =
        {
            id: crypto.randomUUID(),
            level: level,
            category: category,
            title: typeof title === "string" ? title : String(title),
            message: message === undefined || message === null ? "" : String(message),
            service: (options.service !== undefined && options.service !== null) ? options.service : logServiceOrigin.DOCK,
            accountId: options.accountId ? String(options.accountId) : "",
            errorCode: options.errorCode ? String(options.errorCode) : "",
            errorReason: options.errorReason ? String(options.errorReason) : "",
            additionalData: (options.additionalData && typeof options.additionalData === "object") ? options.additionalData : {},
            timestamp: now,
            timestampIsoString: now.toISOString(),
            sequence: Logger.#nextSequence(),
            environment: Logger.#environment
        };

        // Mirror to standard output (journald). To keep production journald lean,
        // only WARNING/ERROR are mirrored unless --debug is on; every entry is
        // still persisted to MongoDB regardless.
        if (Logger.#debugEnabled || level >= logLevel.WARNING)
        {
            try
            {
                process.stdout.write(`${LogFormatter.formatLine(logEntryDocument)}\n`);
            }
            catch (standardOutputError)
            {
                // A closed stdout must never break the app.
            }
        }

        LogIngester.write(logEntryDocument);
    }

    static debug(category, title, message, options = {})
    {
        Logger.#record(logLevel.DEBUG, category, title, message, options);
    }

    static info(category, title, message, options = {})
    {
        Logger.#record(logLevel.INFO, category, title, message, options);
    }

    static warning(category, title, message, options = {})
    {
        Logger.#record(logLevel.WARNING, category, title, message, options);
    }

    static error(category, title, message, options = {})
    {
        Logger.#record(logLevel.ERROR, category, title, message, options);
    }

    /**
     * Records at an explicit numeric level. Used by the /Logs/Ingest endpoint to
     * persist browser-emitted entries (which carry their own level and a WEB
     * service origin via options.service).
     */
    static record(level, category, title, message, options = {})
    {
        Logger.#record(level, category, title, message, options);
    }

    /**
     * Legacy signature. Existing call sites pass (message, sourceTag); map them to
     * a structured SYSTEM/INFO entry with the tag as the title so those logs are
     * captured rather than dropped.
     */
    static log(message, source = "DOCK")
    {
        Logger.#record(logLevel.INFO, logCategory.SYSTEM, String(source), message, {});
    }

    /**
     * Legacy worker-output echo used only in --debug. The Agent now logs itself
     * directly to MongoDB, so this stays a standard-output convenience (no double
     * persistence) for a developer watching a local worker.
     */
    static logWorker(taskTypeName, taskId, stream, line)
    {
        if (!Logger.#debugEnabled)
        {
            return;
        }
        if (line === "" || line == null)
        {
            return;
        }

        const shortId = taskId ? String(taskId).slice(0, 8) : "????????";
        const tag = `AGENT:${taskTypeName}:${shortId}${stream === "stderr" ? ":err" : ""}`;
        try
        {
            process.stdout.write(`[${new Date().toISOString()}] [${tag}] ${line}\n`);
        }
        catch (standardOutputError)
        {
            // ignore
        }
    }
}

module.exports = Logger;

File: Agent/Globals/Classes/Logging/Logger.py
# The Agent's structured logging entry point. Mirrors Dock's Logger: every call
# builds a canonical log entry (with the timestamp stamped INTERNALLY, never a
# caller parameter) and writes it DIRECTLY to the shared logEvents collection —
# the same trust model as CreditLedger, which already writes money this way. This
# is what makes Agent (and burst-virtual-machine) work visible in the central log
# in production, where the old print shim was a no-operation.
#
# Writes are direct rather than buffered so an entry cannot be lost if a
# short-lived worker exits right after logging. If the database write fails, the
# formatted line is emitted to the real standard output (bypassing any print shim)
# so the entry still survives in the worker's captured output.

import os
import sys
import uuid
from datetime import datetime, timezone

from Globals.Classes.Database.DatabaseConnector import DatabaseConnector
from Globals.Constants.DatabaseConstants import DatabaseConstants
from Globals.Enumerations.LogLevel import LogLevel
from Globals.Enumerations.LogServiceOrigin import LogServiceOrigin
from Globals.Classes.Logging.LogFormatter import LogFormatter


class Logger:

    __sequence_counter = 0

    @staticmethod
    def __next_sequence() -> int:
        Logger.__sequence_counter = (Logger.__sequence_counter + 1) % (2 ** 53)
        return Logger.__sequence_counter

    @staticmethod
    def __resolve_environment() -> str:
        environment = os.getenv("MINDMELD_ENVIRONMENT")
        if environment:
            return environment
        return "local" if "--debug" in sys.argv else "production"

    @staticmethod
    def __build_document(level, category, title, message, account_id, error_code, error_reason, additional_data) -> dict:
        now = datetime.now(timezone.utc)
        return {
            "id": str(uuid.uuid4()),
            "level": int(level),
            "category": int(category),
            "title": str(title),
            "message": "" if message is None else str(message),
            "service": int(LogServiceOrigin.AGENT),
            "accountId": str(account_id) if account_id else "",
            "errorCode": str(error_code) if error_code else "",
            "errorReason": str(error_reason) if error_reason else "",
            "additionalData": additional_data if isinstance(additional_data, dict) else {},
            "timestamp": now,
            "timestampIsoString": now.isoformat(),
            "sequence": Logger.__next_sequence(),
            "environment": Logger.__resolve_environment(),
        }

    @staticmethod
    async def __record(level, category, title, message, account_id="", error_code="", error_reason="", additional_data=None) -> None:
        document = Logger.__build_document(level, category, title, message, account_id, error_code, error_reason, additional_data)

        try:
            database = await DatabaseConnector.get_database()
            if database is not None:
                database[DatabaseConstants.LOG_EVENTS_COLLECTION].insert_one(document)
                return
        except Exception:
            pass

        # Fallback: write the formatted line to the real standard output (not the
        # possibly-shimmed print) so the entry is captured rather than lost.
        try:
            sys.__stdout__.write(LogFormatter.format_line(document) + "\n")
            sys.__stdout__.flush()
        except Exception:
            pass

    @staticmethod
    async def debug(category, title, message, account_id="", error_code="", error_reason="", additional_data=None) -> None:
        await Logger.__record(LogLevel.DEBUG, category, title, message, account_id, error_code, error_reason, additional_data)

    @staticmethod
    async def info(category, title, message, account_id="", error_code="", error_reason="", additional_data=None) -> None:
        await Logger.__record(LogLevel.INFO, category, title, message, account_id, error_code, error_reason, additional_data)

    @staticmethod
    async def warning(category, title, message, account_id="", error_code="", error_reason="", additional_data=None) -> None:
        await Logger.__record(LogLevel.WARNING, category, title, message, account_id, error_code, error_reason, additional_data)

    @staticmethod
    async def error(category, title, message, account_id="", error_code="", error_reason="", additional_data=None) -> None:
        await Logger.__record(LogLevel.ERROR, category, title, message, account_id, error_code, error_reason, additional_data)
